### PR TITLE
Fixes #19313 - autoprovisioning NIC association fixed

### PR DESCRIPTION
**Pull request by lzap**
After discovered host is converted to Managed, associated NICs still have their associations set to the original record from the database (`Host::Discovered`), which led the `host.managed?` statement fail in `tftp_ready?` method in TFTP orchestration. This led to TFTP orchestration to be skipped completely, so I added a statement to set the appropriate assotiation explicitly during auto provisioning.

This is a regression and I cannot find the root cause of this, it worked previously. We have reports that both 1.14 and 1.15 are broken.

### CODE REVIEW

**Comment by timogoebel** [comment]
This is a nasty one. Reminds me a lot of e9c626c7258ce405dbe3d95051a5692ac1d87763. Have you tried if a similar approach helps here?

File: test/test_helper_discovery.rb [hunk 1]
```diff
@@ -1,3 +1,7 @@
+root = File.expand_path(File.dirname(__FILE__))
+require "#{root}/factories/discovery_rule_related"
+require "#{root}/factories/discovery_host_related"
```
Comment (lzap) [inline test/test_helper_discovery.rb]: This adds possibility to run individual tests:

    be rake test:discovery TEST=/home/lzap/work/foreman_discovery/test/unit/discovered_extensions_test.rb

File: app/controllers/concerns/foreman/controller/discovered_extensions.rb [hunk 2]
```diff
@@ -57,8 +57,15 @@ def perform_auto_provision original_host, rule
...
     host.attributes = host.apply_inherited_attributes(hostgroup_id: rule.hostgroup_id)
     host.set_hostgroup_defaults
+    # interfaces are associated with old record which prevents from TFTP orchestration
+    host.primary_interface.host = host
     # save! does not work here
-    host.save ? host : false
+    if host.save
+      host
+    else
+      Rails.logger.error "Auto provisioning failed: #{host.errors.full_messages.to_sentence}"
```
Comment (lzap) [inline app/controllers/concerns/foreman/controller/discovered_extensions.rb:64]: This also closes http://projects.theforeman.org/issues/19405

**Comment by lzap** [comment]
Strange error on Jenkins:

```
NoMethodError: undefined method `split' for nil:NilClass
     (NoMethodError)
/var/lib/workspace/workspace/test_plugin_matrix/database/postgresql/ruby/2.0.0/slave/fast/plugin/test/unit/discovered_extensions_test.rb:176
```

**Comment by timogoebel** [comment]
> Strange error on Jenkins:

I believe, this can be fixed by replacing `assert_equal nil, blubb` with `assert_ni blubb`.

**Comment by lzap** [comment]
Adding `refute_nil` there.

**Comment by lzap** [comment]
Indeed this looks like that was the root cause, I did not notice it. Thank to our weak tests we did not catch it at all. I added tests for autoprovisioning which will verify TFTP and DHCP orchestration.

**Comment by lzap** [comment]
Hmmm re-testing this today, it does not seem to fix the issue. If I comment out the `#host.clear_association_cache` then it works. Any idea how to approach this?

**Comment by timogoebel** [comment]
> Any idea how to approach this?

Not at all. I just remember going nuts to come up with this patch.
Maybe we should post to theforeman-dev and ask for help there. Someone might have an idea.

File: test/test_helper_discovery.rb (hunk 1 above)
Comment (timogoebel) [reply test/test_helper_discovery.rb]: Would `require_relative` be an alternative to this?

Comment (lzap) [reply test/test_helper_discovery.rb]: I learned something, thanks. It works yeah.

**Comment by lzap** [comment]
@timogoebel since the regression is fatal and folks cant provision via auto or even API, we might want to remove the line you added for now and try to find better solution later?

To be honest, I don't know why you clear all associations when becomes is overloaded in Host::Base and does exactly that for you:

```ruby
    def becomes(*args)
      became = super
      became.drop_primary_interface_cache
      became.drop_provision_interface_cache
      became.interfaces = self.interfaces
      became
    end
```

**Comment by timogoebel** [comment]
@lzap : I'd be ok with dropping the line. The overloaded becomes method does not help, might even make things worse.

The issue is iirc, that `host != host.primary_interface.host`. Clearing host's cache doesn't help.

**Comment by lzap** [comment]
Ok added more tests including check for subnet.tftp in auto provisioning and fixed factories so they now work with or without `TEST=` option in rake. I cannot reproduce the case you were adding the line for, so I am removing it.

**Comment by lzap** [comment]
I don't understand these, I have `refute_nil` there:

```
NoMethodError: undefined method `split' for nil:NilClass
     (NoMethodError)
/var/lib/workspace/workspace/test_plugin_matrix/database/postgresql/ruby/2.4/slave/fast/plugin/test/unit/discovered_extensions_test.rb:176  
```

**Comment by lzap** [comment]
Any ETA @timogoebel or @tbrisker this is a nasty regression in both 1.14 and 1.15. Thanks!

**Comment by lzap** [comment]
Oh wait tests are still red.

**Comment by timogoebel** [comment]
@lzap : I want to check this out before lunch, hang on...

File: test/factories/discovery_host_related.rb [hunk 3]
```diff
@@ -1,24 +1,26 @@
-FactoryGirl.define do
-  factory :discovered_host, class: 'Host::Discovered' do
-    sequence(:name) { |n| "host#{n}" }
-    sequence(:ip)   { |n| IPAddr.new(n, Socket::AF_INET).to_s }
-    sequence(:mac)  { |n| "02:23:45:67:" + n.to_s(16).rjust(4, '0').insert(2, ':') }
+unless FactoryGirl.factories.registered?(:discovered_host)
```
Comment (timogoebel) [inline test/factories/discovery_host_related.rb]: What is this good for?

Comment (lzap) [reply test/factories/discovery_host_related.rb]: So when you run tests via `TEST=xyz` it loads files in a different way and factory_girl was giving an error - Factory already registered. I could not find any other way than doing this, I want to be able to run all tests and a particular test.

**Comment by lzap** [comment]
Thanks, do you also see a failure like Jenkins?

File: test/unit/discovery_attribute_set_test.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
-require 'test_helper'
```
Comment (timogoebel) [inline test/unit/discovery_attribute_set_test.rb:1]: If we do this, we can remove the file `test/test_helper.rb`. It isn't any good anymore.

**Review by timogoebel — CHANGES_REQUESTED** [review]

File: test/unit/discovery_attribute_set_test.rb (hunk 4 above)
Comment (lzap) [reply test/unit/discovery_attribute_set_test.rb:1]: I actually can't that does not work, this requires `test_helper` from Foreman and not discovery. It's a mess.

Comment (lzap) [reply test/unit/discovery_attribute_set_test.rb:1]: And core `test_helper` required plugins `test_helper`. You need to be this explicit.

Comment (timogoebel) [reply test/unit/discovery_attribute_set_test.rb:1]: Actually, https://github.com/theforeman/foreman_plugin_template doesn't have a `test/test_helper.rb`at all. So this should be possible.

How about doing this in a separate PR to get the actual fix in fast?

**Comment by lzap** [comment]
Incorporated your patch and added refute_nils everywhere.

**Comment by lzap** [comment]
I think I can remove `test_helper.rb` now, let's see what Jenkins does with this version.

**Comment by lzap** [comment]
Might want to split this into two PRs, one is the line change and the other is test improvement.

**Comment by lzap** [comment]
Or just two commits.

**Comment by timogoebel** [comment]
> Or just two commits.

Yep. Makes backporting easier.

What's still missing is the removal of `test/test_helper.rb`

**Comment by lzap** [comment]
Split into two commits, removed the `test_helper.rb` leftover. If it passes locally for you (green for me) let's merge it and if develop is still broken I will create another PR to fix just jenkins tests.

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
Merged, thanks @lzap.

**Pull request merged**